omod/src/main/java/org/openmrs/web/user/CurrentUsers.java

```diff
@@ -14,30 +14,33 @@
 import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
+import java.util.concurrent.CopyOnWriteArrayList;
 
 import javax.servlet.ServletContext;
 import javax.servlet.http.HttpSession;
 
 import org.apache.commons.lang.StringUtils;
-import org.apache.commons.logging.Log;
-import org.apache.commons.logging.LogFactory;
+import org.openmrs.api.context.Context;
 import org.openmrs.User;
 import org.openmrs.web.WebConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Utility methods which maintain the list of current users.
  */
 public class CurrentUsers {
 	
-	private static final Log log = LogFactory.getLog(CurrentUsers.class);
+	private static final Logger log = LoggerFactory.getLogger(CurrentUsers.class);
 	
 	/**
 	 * Initialize the current users list.
 	 * 
 	 * @param servletContext
 	 */
-	public static Map<String, String> init(ServletContext servletContext) {
-		Map<String, String> currentUserMap = Collections.synchronizedMap(new TreeMap<String, String>());
+	public static Map<String, CopyOnWriteArrayList<HttpSession>> init(ServletContext servletContext) {
+		Map<String, CopyOnWriteArrayList<HttpSession>> currentUserMap = Collections
```

Comment:
The main reason why I resorted to `CopyOnWriteArryList` was that I ran into `java.util.ConcurrentModificationException `as I was trying to remove a session from the sessions-list while iterating trough all other existing sessions.  
This is how it all happened.

```private static void invalidateAllUserSessions(String currentUserName,
	.....
			for (HttpSession session : sessions) {
				if (session != null) {
					session.invalidate();
				}
			}
```
which calls the method `removeSessionFromList` through the `SessionListener` class run on every session invalidation. 

@ibacher, Basing on what I have explained above, will option 2 bring out what concurrency very well?

Reply:
That's likely to happen with almost any Java list implementation. You just can't change a list directly while iterating over it. Usually, the way around it it something like this:

```java
Iterator<HttpSession> iterator = sessions.iterator();
while (iterator.hasNext()) {
	HttpSession session = iterator.next();
	session.invalidate();
	iterator.remove();
}
```

The `for (HttpSession session : sessions)` bit is actually just syntactic sugar for the first 3 lines of the above example.

(PS if the `CopyOnWriteArrayList` solved the problem, it's because you're creating a bunch of useless copies).

Reply:
I used the iterator too before I changed to the `CopyOnWriteArryList`


> ```java
> Iterator<HttpSession> iterator = sessions.iterator();
> while (iterator.hasNext()) {
> 	HttpSession session = iterator.next();
> 	session.invalidate();
> 	iterator.remove();
> }
> ```

if we use this, the SessionListener will also try to remove the element on session invalidation.  In my thinking, I expect some errors here :)  

Sometimes sessions are invalidated even when the user hasn't clicked logout. or password change for example when the instance is left unused for like 20 minutes .  
My main goal is to use the listener to update the sessions list each time the session is invalidated to prevent having null objects in the list.  But this takes place from a different method on a different state of the system.

I don't know if this is clear!  But that's how I had understood this.

Reply:
So, the `SessionListener` will never remove any item from the list. This is because it doesn't know about the list and has no way of removing it. So the list should **never** contain any `null` objects.p

What it might make sense to do is to instead have a `pruneExpiredSessions()` method that does something like this:

```java
public static void pruneExpiredSessions(List<HttpSession> sessions) {
	  long currentTimestamp = System.currentTimeMillis();
	  Iterator<HttpSession> iterator = sessions.iterator();
	  while (iterator.hasNext()) {
		  HttpSession session = iterator.next();
		  if (session.getLastAccessTime() - currentTimestamp > session.getMaxInactiveInterval()) {
			  iterator.remove();
		  }
	  }
}
```

Reply:
And, yeah, if you're having an issue with an infinite loop of invalidations, then thing to do is to track which session your actually already invalidating, e.g.,

```java
private static void invalidateAllUserSessions(HttpSession httpSession, String currentUserName,
	        Map<String, CopyOnWriteArrayList<HttpSession>> currentUsers) {
		log.info("Finding sessions for the user: {}", currentUserName);

		CopyOnWriteArrayList<HttpSession> sessions = currentUsers.get(currentUserName);
		if (sessions != null) {
			for (HttpSession session : sessions) {
				if (!httpSession.equals(session)) {
					session.invalidate();
				}
			}

			currentUsers.remove(currentUserName);
			log.debug("Found {} sessions for the user: {}", sessions.size(), currentUserName);
		} else {
			log.debug("No sessions found for this user: {}", currentUserName);
		}
	}
```

Reply:
> And, yeah, if you're having an issue with an infinite loop of invalidations, then thing to do is to track which session your actually already invalidating, e.g.,

Thank you @ibacher for really taking some time to look at this.  I am going to polish this and then push the code.